### PR TITLE
[SPARK-38405][SQL] Supports customization of the entire optimizer and planner

### PR DESCRIPTION
**Pull request by chenjunbiao001**
linked to issue:https://issues.apache.org/jira/browse/SPARK-38405

### CODE REVIEW

**Comment by chenjunbiao001** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，邮件已收到，感谢您的来信！

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala [hunk 1]
```diff
@@ -242,15 +242,24 @@ abstract class BaseSessionStateBuilder(
...
   protected def optimizer: Optimizer = {
-    new SparkOptimizer(catalogManager, catalog, experimentalMethods) {
-      override def earlyScanPushDownRules: Seq[Rule[LogicalPlan]] =
-        super.earlyScanPushDownRules ++ customEarlyScanPushDownRules
+    val optimizerClassName = conf.getConfString("spark.sql.customized.optimizer", "")
+    if (optimizerClassName.nonEmpty && !optimizerClassName.equals("")) {
+      val catalogManagerTypeName = Class.forName("org.apache.spark.sql.connector.catalog.CatalogManager")
+      val sessionCatalogTypeName = Class.forName("org.apache.spark.sql.catalyst.catalog.SessionCatalog")
+      val experimentalMethodsTypeName = Class.forName("org.apache.spark.sql.ExperimentalMethods")
+      val optimizerInstance = Class.forName(optimizerClassName).getConstructor(catalogManagerTypeName, sessionCatalogTypeName, experimentalMethodsTypeName).newInstance(catalogManager, catalog, experimentalMethods)
```
Comment (nyingping) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala:250]: file line length exceeds 100 characters on line 247,248,250 and 304

https://github.com/databricks/scala-style-guide#linelength

Comment (nyingping) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala:250]: file line length exceeds 100 characters on line 247,248,250 and 304

https://github.com/databricks/scala-style-guide#linelength

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala [hunk 2]
```diff
@@ -242,15 +242,24 @@ abstract class BaseSessionStateBuilder(
...
+      optimizerInstance.asInstanceOf[Optimizer]
+    } else {
+      new SparkOptimizer(catalogManager, catalog, experimentalMethods) {
+        override def earlyScanPushDownRules: Seq[Rule[LogicalPlan]] =
+          super.earlyScanPushDownRules ++ customEarlyScanPushDownRules
 
-      override def preCBORules: Seq[Rule[LogicalPlan]] =
-        super.preCBORules ++ customPreCBORules
+        override def preCBORules: Seq[Rule[LogicalPlan]] =
+          super.preCBORules ++ customPreCBORules
```
Comment (nyingping) [inline sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala:258]: nit. redundant space on line 257,258.

**Comment by nyingping** [comment]
1.PR title should be changed to: `[SPARK-38405][SQL] Supports customization of the entire optimizer and planner`

2.PR description may be in the default format to make the community better understand your intention.

```
### What changes were proposed in this pull request?

### Why are the changes needed?

### Does this PR introduce any user-facing change?

### How was this patch tested?
```
3.You should open GA in your fork repository

https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository

File: sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala (hunk 1 above)
Comment (wangyum) [reply sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala:250]: @chenjunbiao001 @nyingping Could we use `spark.sql.extensions` to to add custom optimizer rules?

**Comment by chenjunbiao001** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，邮件已收到，感谢您的来信！

**Pull request closed without merging**